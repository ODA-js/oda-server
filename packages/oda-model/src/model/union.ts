import {
  MetaModelType,
  UnionStorage,
  UnionInput,
  UnionMeta,
} from './interfaces';
import { ModelBase } from './modelbase';
import { IUnion } from './interfaces/model';

export class Union extends ModelBase<UnionMeta, UnionInput, UnionStorage>
  implements IUnion {
  public modelType: MetaModelType = 'union';
  constructor(inp: UnionInput) {
    super(inp);
  }
  get items(): Set<string> {
    return this.$obj.items;
  }

  public updateWith(obj: UnionInput) {
    super.updateWith(obj);
    this.$obj.items = new Set(obj.items);
  }

  public toObject(): UnionInput {
    return {
      ...super.toObject(),
      items: [...this.items],
    };
  }
}
