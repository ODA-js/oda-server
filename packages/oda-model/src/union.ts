import { merge } from 'lodash';
import { ElementMetaInfo } from './element';
import {
  ModelBaseInternal,
  ModelBaseInput,
  ModelBase,
  IModelBase,
  ModelBaseOutput,
} from './modelbase';
import { MetaModelType, Nullable, assignValue } from './types';

/**
 * union definition
 */
export interface IUnion
  extends IModelBase<UnionMetaInfo, UnionInput, UnionOutput> {
  /**
   * item list
   */
  readonly items: Set<string>;
}

export interface UnionMetaInfo extends ElementMetaInfo {}

export interface UnionInternal extends ModelBaseInternal<UnionMetaInfo> {
  items: Set<string>;
}

export interface UnionInput extends ModelBaseInput<UnionMetaInfo> {
  items: string[];
}

export interface UnionOutput extends ModelBaseOutput<UnionMetaInfo> {
  items: string[];
}

const defaultMetaInfo = {};
const defaultInternal = {};
const defaultInput = {};

export class Union
  extends ModelBase<UnionMetaInfo, UnionInput, UnionInternal, UnionOutput>
  implements IUnion {
  public modelType: MetaModelType = 'union';
  constructor(inp: UnionInput) {
    super(merge({}, defaultInput, inp));
    this.metadata_ = merge({}, defaultMetaInfo, this.metadata_);
    this.$obj = merge({}, defaultInternal, this.$obj);
  }
  get items(): Set<string> {
    return this.$obj.items;
  }

  public updateWith(input: Nullable<UnionInput>) {
    super.updateWith(input);
    assignValue<UnionInternal, UnionInput, UnionInput['items']>({
      src: this.$obj,
      input,
      field: 'items',
      effect: (src, value) => (src.items = new Set(value)),
      required: true,
      setDefault: src => (src.items = new Set()),
    });
  }

  public toObject(): UnionOutput {
    return merge({}, super.toObject(), { items: [...this.items] } as Partial<
      UnionOutput
    >);
  }
}
