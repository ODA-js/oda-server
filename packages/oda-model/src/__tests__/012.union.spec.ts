import 'jest';
import { Union, unionDefaultMetaInfo } from '../union';

describe('Union', () => {
  it('default', () => {
    const res = new Union({
      name: 'UNION',
      items: ['ITEM', 'Item', 'item'],
    });
    expect(res.metadata).toMatchObject(unionDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('dedupes', () => {
    const res = new Union({
      name: 'UNION',
      items: ['ITEM', 'Item', 'item', 'item', 'item'],
    });
    expect(res.metadata).toMatchObject(unionDefaultMetaInfo);
    expect(res.items.size).toBe(3);
  });
  it('merges', () => {
    const res1 = new Union({
      name: 'UNION',
      items: ['ITEM', 'Item2'],
    });
    const res2 = new Union({
      name: 'UNION',
      items: ['ITEM', 'Item1'],
    });

    const res = new Union({
      name: 'UNION',
      items: ['ITEM', 'Item'],
    });
    res.mergeWith(res1.toObject());
    res.mergeWith(res2.toObject());
    expect(res.items.size).toBe(4);
    expect(res.toObject()).toMatchSnapshot('merge');
  });
});
