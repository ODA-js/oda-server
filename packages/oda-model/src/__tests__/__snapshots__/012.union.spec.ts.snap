// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Union default: toObject 1`] = `
Object {
  "description": undefined,
  "items": Array [
    "ITEM",
    "Item",
    "item",
  ],
  "metadata": Object {},
  "name": "UNION",
  "title": undefined,
}
`;

exports[`Union merges: merge 1`] = `
Object {
  "description": undefined,
  "items": Array [
    "ITEM",
    "Item",
    "Item2",
    "Item1",
  ],
  "metadata": Object {},
  "name": "UNION",
  "title": undefined,
}
`;
